hpe/dl380/exporter.go

```diff
@@ -0,0 +1,553 @@
+	NVME = "NVMeDriveMetrics"
+	// DISKDRIVE represents the Disk Drive metric endpoints
+	DISKDRIVE = "DiskDriveMetrics"
+	// LOGICALDRIVE represents the Logical drive metric endpoint
+	LOGICALDRIVE = "LogicalDriveMetrics"
+	// MEMORY represents the memory metric endpoints
+	MEMORY = "MemoryMetrics"
+	// OK is a string representation of the float 1.0 for device status
+	OK = 1.0
+	// BAD is a string representation of the float 0.0 for device status
+	BAD = 0.0
+	// DISABLED is a string representation of the float -1.0 for device status
+	DISABLED = -1.0
+)
+
+var (
+	log *zap.Logger
+)
+
+// Exporter collects chassis manager stats from the given URI and exports them using
+// the prometheus metrics package.
+type Exporter struct {
+	ctx   context.Context
+	mutex sync.RWMutex
+	pool  *pool.Pool
+	host  string
+
+	up            prometheus.Gauge
+	deviceMetrics *map[string]*metrics
+}
+
+// NewExporter returns an initialized Exporter for HPE DL380 device.
+func NewExporter(ctx context.Context, target, uri string) *Exporter {
+	var fqdn *url.URL
+	var tasks []*pool.Task
+
+	log = zap.L()
+
+	tr := &http.Transport{
+		Dial: (&net.Dialer{
+			Timeout: 3 * time.Second,
+		}).Dial,
+		MaxIdleConns:          1,
+		MaxConnsPerHost:       1,
+		MaxIdleConnsPerHost:   1,
+		IdleConnTimeout:       90 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
+		},
+		TLSHandshakeTimeout: 10 * time.Second,
+	}
+
+	retryClient := retryablehttp.NewClient()
+	retryClient.CheckRetry = retryablehttp.ErrorPropagatedRetryPolicy
+	retryClient.HTTPClient.Transport = tr
+	retryClient.HTTPClient.Timeout = 30 * time.Second
+	retryClient.Logger = nil
+	retryClient.RetryWaitMin = 2 * time.Second
+	retryClient.RetryWaitMax = 2 * time.Second
+	retryClient.RetryMax = 2
+	retryClient.RequestLogHook = func(l retryablehttp.Logger, r *http.Request, i int) {
+		retryCount := i
+		if retryCount > 0 {
+			log.Error("api call "+r.URL.String()+" failed, retry #"+strconv.Itoa(retryCount), zap.Any("trace_id", ctx.Value("traceID")))
+		}
+	}
+
+	// Check that the target passed in has http:// or https:// prefixed
+	fqdn, err := url.ParseRequestURI(target)
+	if err != nil {
+		fqdn = &url.URL{
+			Scheme: config.GetConfig().OOBScheme,
+			Host:   target,
+		}
+	}
+
+	// vars for drive parsing
+	var (
+		initialURL        = "/Systems/1/SmartStorage/ArrayControllers"
+		url               = initialURL
+		chassis_url       = "/Chassis/1"
+		logicalDriveURLs  []string
+		physicalDriveURLs []string
+		nvmeDriveURLs     []string
+	)
+
+	// PARSING DRIVE ENDPOINTS
+	// Get initial JSON return of /redfish/v1/Systems/1/SmartStorage/ArrayControllers/ set to output
+	output, err := getDriveEndpoint(fqdn.String()+uri+url, target, retryClient)
+
+	// Loop through Members to get ArrayController URLs
+	if err != nil {
+		log.Error("api call "+fqdn.String()+uri+url+" failed.", zap.Any("trace_id", ctx.Value("traceID")))
```

Comment:
When logging an error message, if `err` is available please include it in `log.Error()` by using `zap.Error(err)`, please apply it for all instances in this file.